### ISSUE TITLE
Examples

### ISSUE THREAD

**Issue by nerusnayleinad** [containous/maesh#267]
Is there a single example one can start with?

This is everything you have released for usage?

Usage¶
To use maesh, instead of referencing services via their normal `<servicename>.<namespace>`, instead use `<servicename>.<namespace>.maesh`. This will access the maesh service mesh, and will allow you to route requests through maesh.

By default, maesh is opt-in, meaning you have to use the maesh service names to access the mesh, so you can have some services running through the mesh, and some services not.

`referencing services via their normal <servicename>.<namespace>?`. I don't get it.

I'd like to be a hello world to know what's the dynamics.

**Comment by SantoDE** [comment]
Hey! :) 

Let me try to clarify that. By default, no service-to-service communication is going through maesh. To access a service through maesh, you have to call it with the dns `<servicename>.<namespace>.maesh`.

So lets say, your service calls the api of another service. Currently, it takes the address of that api as <servicename.namespace>. If you now want to make that call through maesh, you have to set the base address to `<servicename>.<namespace>.maesh`.

Does that make things more clear?

**Comment by nerusnayleinad** [comment]
Oh. Sorry, I get it now.

Is it straight forward? I am getting 404.

```
/ # curl nginx -I
HTTP/1.1 200 OK
Server: nginx/1.17.4
Date: Tue, 08 Oct 2019 15:09:02 GMT
Content-Type: text/html
Content-Length: 612
Last-Modified: Tue, 24 Sep 2019 14:49:10 GMT
Connection: keep-alive
ETag: "5d8a2ce6-264"
Accept-Ranges: bytes

/ # curl nginx.default -I
HTTP/1.1 200 OK
Server: nginx/1.17.4
Date: Tue, 08 Oct 2019 15:09:04 GMT
Content-Type: text/html
Content-Length: 612
Last-Modified: Tue, 24 Sep 2019 14:49:10 GMT
Connection: keep-alive
ETag: "5d8a2ce6-264"
Accept-Ranges: bytes

/ # curl nginx.default.maesh -I
HTTP/1.1 404 Not Found
Content-Type: text/plain; charset=utf-8
X-Content-Type-Options: nosniff
Date: Tue, 08 Oct 2019 15:09:07 GMT
Content-Length: 19
```

**Comment by ADefossez-Zenika** [comment]
Hello, It should be ... 
Can you send us a "kubectl --namespace=maesh get pods" ? Is everything OK on the Maesh deployment ?

**Comment by nerusnayleinad** [comment]
No. Actually it's not. Everything is running, but when I get the logs, there are some errors. But it's from jaeger, so I thought may be it is not relevant.

Another thing to point out is that when I installed `maesh´ through helm, jaeger and prometheus pods were on pending as they were waiting for a pv to me mounted, but there was no storage class. I installed the `local-path` storage class, which fixed the issue. I believe it is fine, as that storage class just provisions volumes from the host.

These is what's going on in the cluster:

```
# kubectl get po -n maesh
NAME                                READY   STATUS    RESTARTS   AGE
grafana-core-77fddd88fd-nwcqj       1/1     Running   0          7m15s
jaeger-7d9fd6c68d-qfwrx             1/1     Running   0          7m15s
maesh-controller-564bfcdb9f-fkcm5   1/1     Running   0          7m15s
maesh-mesh-8r2jg                    1/1     Running   0          7m6s
prometheus-core-7bb6587788-mdwvw    2/2     Running   0          7m15s
```
This is the errors from maesh-mesh:

`time="2019-10-08T15:37:44Z" level=error msg="Tracing jaeger error: error when flushing the buffer: write udp 127.0.0.1:55501->127.0.0.1:6831: write: connection refused" tracingProviderName=jaeger`

**Comment by SantoDE** [comment]
Can you try to reinstall with setting metrics and tracing to false with the helm install? Just to see if that has an inlfuence?

**Comment by nerusnayleinad** [comment]
I removed the parameters directly from the daemonSet, and yes, it worked. But, as you can see it is not getting the right values during the deployment. If I edit the daemonSet and set the following values, it works perfectly:

```
- --tracing.jaeger.localagenthostport=jaeger-agent.maesh.svc.cluster.local:6831
- --tracing.jaeger.samplingserverurl=jaeger-agent.maesh.svc.cluster.local:5778/sampling
```
<maesh> is the namespace where maesh runs, of course.

**Comment by riker09** [comment]
@nerusnayleinad Can you please give a full example of what commands you have executed to get Maesh up and running? I'm having a hard time getting it going on a vanilla Kubernetes cluster (I'm using Docker on Windows). This *might* be related to the fact that I'm completely new to Kubernetes and I'm still learning. But a full example would really help and would be much appreciated.

**Comment by riker09** [comment]
For the record, I had success with this command:

```bash
helm install \
  --name=maesh \
  --namespace=maesh \
  --set metrics.prometheus.enabled=false \
  --set tracing.jeeger.enabled=false \
  --set metrics.deploy=false \
  --set tracing.deploy=false \
  maesh/maesh
```

I had to disable the tracing and metrics feature as there is some Persistent Volume Claim error that I don't understand (yet).

**Comment by riker09** [comment]
@nerusnayleinad How did you create the PersistantVolume? Are you using minikube?

**Comment by nerusnayleinad** [comment]
@riker09 I don't remember exactly the commands I used, but I manually edited jaeger and prometheus services and it worked fine. In the beginning, yes, the problem was with volumes.

From what I posted above:

Another thing to point out is that when I installed maesh´ through helm, jaeger and prometheus pods were on pending as they were waiting for a PV to me mounted, but there was no storage class. I installed `local-path` storage class, which fixed the issue. I believe it is fine, as that storage class just provisions volumes from the host. 

But my main problem was the 404, which got resolves automagically.

**Comment by riker09** [comment]
I'm having trouble installing mentioned storage class. Here's what I have done:

`local-storage.class.yml`:

```yaml
apiVersion: storage.k8s.io/v1
kind: StorageClass
metadata:
  name: local-path
provisioner: kubernetes.io/no-provisioner
volumeBindingMode: WaitForFirstConsumer
```

Install storage class:

```bash
kubectl create -f local-storage.class.yml
```

Persistant volume definition in `metrics-storage.yml`:

```yaml
apiVersion: v1
kind: PersistentVolume
metadata:
  name: metrics-storage
spec:
  capacity:
    storage: 1Gi
  # volumeMode field requires BlockVolume Alpha feature gate to be enabled.
  volumeMode: Filesystem
  accessModes:
  - ReadWriteOnce
  persistentVolumeReclaimPolicy: Delete
  storageClassName: local-path
  hostPath:
    path: /data/pv0001
  nodeAffinity:
    required:
      nodeSelectorTerms:
      - matchExpressions:
        - key: kubernetes.io/hostname
          operator: In
          values:
          - minikube
```

Create persistant volume:

```bash
kubectl create -f metrics-storage.yml 
```

Here are some of the results I get:

```
kubectl -n maesh describe persistentvolumeclaims 
Name:          metrics-storage
Namespace:     maesh
StorageClass:  local-path
Status:        Pending
Volume:        
Labels:        chart=metrics-0.0.6
               heritage=Tiller
               release=maesh
Annotations:   <none>
Finalizers:    [kubernetes.io/pvc-protection]
Capacity:      
Access Modes:  
VolumeMode:    Filesystem
Mounted By:    grafana-core-77fddd88fd-gvrjv
Events:
  Type    Reason                Age                  From                         Message
  ----    ------                ----                 ----                         -------
  Normal  WaitForFirstConsumer  8s (x10 over 2m18s)  persistentvolume-controller  waiting for first consumer to be created before binding


Name:          prometheus-storage
Namespace:     maesh
StorageClass:  local-path
Status:        Pending
Volume:        
Labels:        chart=metrics-0.0.6
               heritage=Tiller
               release=maesh
Annotations:   <none>
Finalizers:    [kubernetes.io/pvc-protection]
Capacity:      
Access Modes:  
VolumeMode:    Filesystem
Mounted By:    prometheus-core-7bb6587788-svhfr
Events:
  Type    Reason                Age                  From                         Message
  ----    ------                ----                 ----                         -------
  Normal  WaitForFirstConsumer  8s (x10 over 2m18s)  persistentvolume-controller  waiting for first consumer to be created before binding
```

```
ubectl -n maesh describe pod grafana-core-77fddd88fd-gvrjv 
Name:           grafana-core-77fddd88fd-gvrjv
Namespace:      maesh
Priority:       0
Node:           <none>
Labels:         app=grafana
                component=core
                pod-template-hash=77fddd88fd
Annotations:    <none>
Status:         Pending
IP:             
IPs:            <none>
Controlled By:  ReplicaSet/grafana-core-77fddd88fd
Containers:
  grafana-core:
    Image:      grafana/grafana:6.2.5
    Port:       3000/TCP
    Host Port:  0/TCP
    Limits:
      cpu:     100m
      memory:  100Mi
    Requests:
      cpu:        100m
      memory:     100Mi
    Liveness:     http-get http://:web/api/health delay=5s timeout=1s period=10s #success=1 #failure=3
    Readiness:    http-get http://:web/api/health delay=5s timeout=1s period=10s #success=1 #failure=3
    Environment:  <none>
    Mounts:
      /etc/grafana from grafana-config (ro)
      /var/lib/grafana from metrics-storage (rw)
      /var/lib/grafana/dashboards/traefik.json from dashboards-grafana-traefik (rw,path="traefik.json")
Conditions:
  Type           Status
  PodScheduled   False 
Volumes:
  metrics-storage:
    Type:       PersistentVolumeClaim (a reference to a PersistentVolumeClaim in the same namespace)
    ClaimName:  metrics-storage
    ReadOnly:   false
  dashboards-grafana-traefik:
    Type:      ConfigMap (a volume populated by a ConfigMap)
    Name:      grafana-dashboard-traefik
    Optional:  false
  grafana-config:
    Type:        ConfigMap (a volume populated by a ConfigMap)
    Name:        grafana-config
    Optional:    false
QoS Class:       Guaranteed
Node-Selectors:  <none>
Tolerations:     node.kubernetes.io/not-ready:NoExecute for 300s
                 node.kubernetes.io/unreachable:NoExecute for 300s
Events:
  Type     Reason            Age                   From               Message
  ----     ------            ----                  ----               -------
  Warning  FailedScheduling  30s (x11 over 9m55s)  default-scheduler  0/1 nodes are available: 1 node(s) didn't find available persistent volumes to bind.
```

I'm using minikube v1.4.0 under Ubuntu 19.10 with kvm (if that matters for some reason).

From what I understand minikube has support for `hostPath` which is why I have changed the persistent volume definition from `local` to `hostPath` in line `14` of `metrics-storage.yml`. But the outcome is the same. I tried searching for the error methods online but judging from the articles and tutorials I have found everything should be running smoothly... Perhaps you can shine some light on the persistant volume issue?

**Comment by riker09** [comment]
I have created a pod with a pvc in the default namespace, added a pv in the same namespace (still default) and it works? Now I'm severely confused. Persistent Volumes are *not* namespaced, are they?

**Comment by dtomcej** [comment]
PVs are **not** namespaced.
PVCs **are** namespaced.

You need to have a provisioner that is able to dynamically provision storage. This is why the default storage class is required, as many clusters will have a variety of storage available, and the default tells the provisioner which to use.

Many users across many projects have had issues with minikube over the years due to non-conformance with kubernetes standards.

We would recommend using k3s instead, as it is full kubernetes, and fully conformant.

We use k3s in our integration tests, so you can ensure that there is full operability.

**Comment by riker09** [comment]
Thank you, @dtomcej for the clarification on the storage class. I think I figured out where my issue lies now. Feel free to correct me, if I'm wrong but:

```bash
$ kubectl get storageclasses.storage.k8s.io 
NAME                   PROVISIONER                    AGE
local-path (default)   kubernetes.io/no-provisioner   19h
standard               k8s.io/minikube-hostpath       23m
```
Note that I have changed my default storage class from `standard` to `local-path` which I have created earlier. This storage class does not feature a provisioner and is thus not suitable for the PVCs created by the maesh chart, correct?

I'm brand new to this whole kubernetes universe. It would be super-helpful if all those Getting Started guides would actually work. I'm having a very hard time on putting all puzzle pieces together. A hint in the documention "Attention: This does not work on minikube!" would have given me a heads up. :man_shrugging: 

I'll try out `k3s`, thanks for the tip.

**Comment by nerusnayleinad** [comment]
Hi,

Sorry, I missed your message. I used the local path storage class from [this](https://github.com/rancher/local-path-provisioner) repo; from rancher. The provisioner is rancher, and it worked perfectly for me.

**Comment by riker09** [comment]
Thank you very much, @nerusnayleinad for the link to the storage class provider. I have installed it and was finally able to successfully deploy the `maesh` helm chart. I couldn't use `v0.8.0` from the official maesh chart repository, however.

```
Error: release maesh failed: Deployment.apps "grafana-core" is invalid: [spec.selector: Required value, spec.template.metadata.labels: Invalid value: map[string]string{"app":"grafana", "component":"core"}: `selector` does not match template `labels`]
```

This is due to an error that was fixed in #305 . So I simply cloned this repository and installed the chart "by hand".

```
git clone https://github.com/containous/maesh.git
cd maesh/helm/chart/maesh
helm install --name maesh --namespace maesh .
```

I know this is all a little bit off-topic. But I hope this will make it into the examples so the next poor soul that stumbles across this will not be as lost as I was. :)
